### PR TITLE
ARROW-16325: [R] Add task for R package with gcc12

### PR DESCRIPTION
**Pull request by paleolimbot**

### CODE REVIEW

File: dev/tasks/tasks.yml [hunk 1]
```diff
@@ -1321,13 +1321,13 @@ tasks:
...
   # This also has -flto=auto
-  test-r-gcc-11:
+  test-r-gcc-12:
     ci: github
     template: docker-tests/github.linux.yml
     params:
       env:
         UBUNTU: 22.04
-        GCC_VERSION: 11
+        GCC_VERSION: 12
```
Comment (jonkeane) [inline dev/tasks/tasks.yml:1330]: Do we know if we have gcc 11 coverage in a different build? I can look and see if you don't off the top of your head. But if we _don't_ we might want to duplicate these and do both 11 and 12.

File: dev/tasks/tasks.yml [hunk 2]
```diff
@@ -1321,13 +1321,13 @@ tasks:
...
       # S3 support is not buildable with gcc11 right now
       flags: '-e ARROW_S3=OFF'
```
Comment (jonkeane) [inline dev/tasks/tasks.yml:1332]: Would you mind also trying to drop this line and see if it builds? I suspect it won't, but it's possible that the aws sdk + gcc build issues have been resolved by now (and we bumped the sdk version enough that we don't run into the old ones)

**Review by jonkeane — APPROVED** [review]
This is fantastic, thanks! One question + one suggestion of a thing to try that might give us slightly more coverage, though we are OK from a CRAN perspective with this info!

**Review by nealrichardson — APPROVED** [review]
+1, you can see gcc 12 here: https://github.com/ursacomputing/crossbow/runs/6164317428?check_suite_focus=true#step:6:51

I feel better submitting to CRAN now that this has been done.

If you wanted to upstream this to rhub, note that CRAN runs GCC 12 in the debian-gcc-devel build only; the fedora devel build and debian release and patched are still using GCC 11: https://cran.r-project.org/web/checks/check_flavors.html. This job runs on ubuntu so it's not identical.

**Review by jonkeane — APPROVED** [review]
OOOH nice to see that S3 works too!

I'm going to go ahead and merge this, but will also [create a jira for upstreaming this change to rhub + have us use that when it's ready.](https://issues.apache.org/jira/browse/ARROW-16388)

**Pull request closed without merging**